### ISSUE TITLE
Allow use of prior studies (have/have not participated) in eligibility criteria

### ISSUE THREAD

**Issue by kimberscott** [lookit/lookit-api#625]
# TL;DR
Allow specifying that to participate in a given study you must have participated in a certain study/studies or must not have participated in a certain study/studies

# Narrative
As a researcher, I may run follow-up studies where I need to ensure samples do not overlap (e.g. an internal replication) or longitudinal follow-up studies.

# Acceptance Criteria

- [ ] It's possible to specify that someone must not have participated in any of XYZ studies using the study UUIDs, either in the criteria expression or in a separate field. For internal studies, this excludes if you participated and had a record with completed_consent_frame.  For external studies, this excludes if you have participated/expressed interest, without checking completed_consent_frame.
- [ ] It's possible to specify that someone must have participated in all of XYZ studies using the study UUIDs, either in the criteria expression or in a separate field. This excludes if you didn't participate with completed_consent_frame. For external studies, this excludes if you haven't participated (without checking  for completed_consent_frame.)
- [ ] Tests of eligiblity checks include both types of criteria

# Implementation Notes

## Kim note

It would be *straightforward* to add fields like participated_<study_UUID> in _get_expanded_child for use in the criteria expression, but the challenge will be figuring out if the performance impact is acceptable and if there are more efficient ways to do this. It's hard to avoid having to retrieve records of this child's responses when checking eligibility, which will be a lot e.g. when sending announcement emails. 

## Which studies can a researcher condition on? 

Goal: Any study they are allowed to read the description/'more details' page of. These categories can be seen as the union of two existing views:

1) `https://lookit.mit.edu/studies/` with no filters selected.  When I view this, there are 4 pages at the bottom; the set should include all studies, not just ones on the first page.

2) `https://lookit.mit.edu/exp/studies/?state=all` Again, we want the full list, not just the first page (so, all 2 pages for my test researcher account, all 102 studies for my admin account)

## How do researchers select/input studies?

Represent this in the study creation/edit forms as two fields:

* Previous Studies - Required

* Previous Studies - Exclude

Q for @okaycj - what will be the best way to represent these 2 lists of studies on the back end?

A: One-to-many table for each.  Each table will have three columns: id, study, previous study.  Here's what schema might look like.

| PreviousStudiesRequired | PreviousStudiesExcluded | 
| ------------------------- | ------------------------- |
| id - primary key, int | id - primary key, int |
| study - foreign key, int | study - foreign key, int |
| previous_study - foreign key, int | previous_study - foreign key, int |

Next, there are two points in the code that will need to be updated.  The SQL query that generates email targets.  This can be found in `studies/tasks.py`.  This SQL query needs to be refactored into Django's ORM, because it is off pattern.  These two tables then can be used as filters.

The second change is with child eligibility.  This can be found in `accounts/queryies.py`.  We currently only use criteria expression to determine eligibility.  The package that is used to "decode" the expression breaks on update (#816) and should be resolved.  Once the tables are build and expression parser is worked out, adding the two tables to the function as filters is pretty straight forward.  

It might be nice to determine child eligibility before emailing the participant.  Additionally, tracking who've we've emailed about about which study, could inform future emails.  

(End of answer to above question.)

It would be okay for researchers to have to select studies by pasting in the long uuid code, but they should be able to see e.g. the names of the studies that are being ruled in or out so they know they selected the correct studies

**Comment by jgerard417** [comment]
I have to have two separate Lookit studies for counterbalancing and want to make sure I don't email people who can't participate in the second. This function would be helpful to stop emails from going out to parents in one condition who have already participated in the other condition.
As one of the conditions is an internal study and one is an external study, it would be preferred if this functionality was the same in both types of studies. The study is set up this way because it has an asynchronous component and a Zoom session; in one condition parents complete the asynchronous part on Lookit and then sign up for the Zoom session (internal, discoverable), while in the other condition parents sign up for the Zoom session (external, discoverable) and are then redirected to the asynchronous part (internal, not discoverable)

**Comment by mekline** [comment]
@okaycj Let's make sure to include @jgerard417 's use case - a paired set of internal and external studies, where participating in one makes you ineligible for the other.
